contributors/design-proposals/watchable-discovery.md

```diff
@@ -0,0 +1,231 @@
+- only showing those GroupVersions that are registered via an
+  `APIService`
+- only showing those GroupVersions that are accepted by the aggregator
+  (i.e. which have no naming conflict).
+
+The aggregator will update its own discovery API group resources
+triggered by changes in the sub-API-servers.
+
+### Embedded Aggregator in kube-apiserver
+
+In this mode, both the aggregator and the kube-apiserver share the
+storage. To avoid conflicts about updating the API group, the `Group`
+objects of kube-apiserver are marked as "local" with an
+annotation. The aggregator will not touch them and the other way
+around for the kube-apiserver for the not-marked `Group` objects.
+
+### HA
+
+API servers in HA mode will share a storage. Hence, they will share
+the state of the discovery API group. Their controllers will fight for
+the correct state of the discovery information.
+
+During normal operation the desired discovery state is equal in an HA
+setup. So no real race will take place.
+
+During rolling updates of the control plane, an old and a new API
+server version might be active. Their reconciliation loops will race
+to update the stored groups.
+
+The consumers of the discovery information might notice the
+flipping. But, this is not a regression of the cluster behaviour as it
+already happens now if a load-balancer distributes requests among the
+cluster API servers, each returning different discovery information.
+
+A long reconciliation period will reduce the effect of the flipping,
+e.g. an API server update shouldn't take longer than a minute.
+
+We could introduce a server version for each API group. Then an older
+server could be less aggressive about reconciliation of new
+groups. But, this is not part of the API types yet.
+
+### Alternative Architectures
+
+- We could implement the discovery resource in the aggregator only and
+  use polling to the sub-API-servers. This would look mostly the same
+  for the cluster end-user, but would be less real-time. Moreover,
+  speaking to the sub-API-servers directly without going through the
+  aggregator would show a different behavior which is not desirable.
+- We could implement an in-memory variant of the discovery
+  resource. While this is considerable more complex (painful
+  experience exists in the OpenShift project), we still have to
+  implement the always-ingreasing behavior of the `ResourceVersion`
+  field. In a distributed environment, this requires some kind of
+  state. This state would naturally be provided by etcd. At this
+  points, it's not that far from storing the `Group` objects
+  persistently in etcd and inherit all the machinery. E.g. we can use
+  normal informers to watch existing API groups, versions and
+  resources.
+- We could extend the current endpoints to be watchable (the main goal
+  of this proposal). But this would mean to reinvent a lot of
+  infrastructure we already have.
+- We could re-use the `meta.k8s.io` group name instead of
```

Comment:
I don't want to add things to meta.k8s.io that aren't about raw API server mechanics.  Discovery is a layer on top of that.